test/unit/org/apache/cassandra/utils/CassandraGenerators.java

```diff
@@ -825,4 +876,145 @@ public static Gen<DecoratedKey> decoratedKeys(Gen<IPartitioner> partitionerGen,
     {
         return rs -> partitionerGen.generate(rs).decorateKey(keyGen.generate(rs));
     }
+
+    public static String insertCqlAllColumns(TableMetadata metadata)
+    {
+        StringBuilder sb = new StringBuilder();
+        sb.append("INSERT INTO ").append(metadata.toString()).append(" (");
+        Iterator<ColumnMetadata> cols = metadata.allColumnsInSelectOrder();
+        while (cols.hasNext())
+            sb.append(cols.next().name.toCQLString()).append(", ");
+        sb.setLength(sb.length() - 2); // remove last ", "
+        sb.append(") VALUES (");
+        for (int i = 0; i < metadata.columns().size(); i++)
+        {
+            if (i > 0)
+                sb.append(", ");
+            sb.append('?');
+        }
+        sb.append(')');
+        return sb.toString();
+    }
+
+    public static void visitUDTs(TableMetadata metadata, Consumer<UserType> fn)
+    {
+        Set<UserType> udts = CassandraGenerators.extractUDTs(metadata);
+        if (!udts.isEmpty())
+        {
+            Deque<UserType> pending = new ArrayDeque<>(udts);
+            Set<ByteBuffer> visited = new HashSet<>();
+            while (!pending.isEmpty())
+            {
+                UserType next = pending.poll();
+                Set<UserType> subTypes = AbstractTypeGenerators.extractUDTs(next);
+                subTypes.remove(next); // it includes self
+                if (subTypes.isEmpty() || subTypes.stream().allMatch(t -> visited.contains(t.name)))
+                {
+                    fn.accept(next);
+                    visited.add(next.name);
+                }
+                else
+                {
+                    pending.add(next);
+                }
+            }
+        }
+    }
+
+    public static class DataGeneratorBuilder
```

Comment:
I prob should migrate all usage to the `ast.Statement`, this was kinda a limited hack and switching to the ast made things far better covered

Reply:
this is actually used in `RandomSchemaTest`.. the reason for the refactor is no longer needed, but I still prefer this...